src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -1275,6 +1275,15 @@ private async ValueTask SendDataAsync(ReadOnlyMemory<byte> buffer, CancellationT
                         await _connection.SendStreamDataAsync(StreamId, current, flush, _requestBodyCancellationSource.Token).ConfigureAwait(false);
                     }
                 }
+                catch (Exception)
+                {
+                    if (_resetException is Exception resetException)
```

Comment:
Should we check and throw the exception the same way as `CheckResponseBodyState` does in:
https://github.com/dotnet/runtime/blob/56778f0b3bc724fe4dae87e626ebc1403db99bb1/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs#L923-L931
`ProcessGoAwayFrame` sets `_canRetry` to `true` AFAICT and we're not handling it here.

Could we just directly call `CheckResponseBodyState` in the exception handling block?

Reply:
`_canRetry` check has been added, but we cannot call `CheckResponseBodyState` here because it asserts for a lock like this
```Debug.Assert(Monitor.IsEntered(SyncObject));```

Reply:
Should we just take the lock and call it, then?

CheckResponseBodyState also checks for `_responseProtocolState == ResponseProtocolState.Aborted`; should we be checking that here too?

Reply:
I don't think we should **not** check `_responseProtocolState` here because `SendDataAsync` is called to send request data, so I believe it should control request-related logic only. That's another reason why I think we shouldn't call `CheckResponseBodyState` here.

Reply:
@geoffkizer Could you please check out my reply above?

Reply:
Ok, your comment sounds right to me.